qiime2/core/validate.py

```diff
@@ -0,0 +1,105 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2021, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from qiime2.core.exceptions import ValidationError, ImplementationError
+from qiime2.core.transform import ModelType
+from qiime2.core.util import sorted_poset
+
+
+class ValidationObject:
+    def __init__(self, concrete_type):
+        self._validators = []
+        self.concrete_type = concrete_type
+        self._is_sorted = False
+
+    def add_validator(self, validator_record):
+        """
+        Used by Plugin to add a `ValidatorRecord` for a new validator to a
+        plugin. See plugin/plugin for definition of `ValidatorRecord`. Usually
+        called through the `register_validtor` decorator.
```

Comment:
These aren't quite numpydoc docstrings. I would check out scikit-bio or qiime2/sdk/usage.py for some good examples.